### PR TITLE
Initial hyperv implementation

### PR DESCRIPTION
**Pull request by mkhon**

### CODE REVIEW

File: src/hyperv/vmbus/vmbus_xact.c [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+    u64 flags = spin_lock_irq(&ctx->xc_lock);
+    /*
+     * NOTE:
+     * xc_active could be NULL, if the ctx has been orphaned.
+     */
+    if (ctx->xc_active != NULL) {
+        vmbus_xact_save_resp(ctx->xc_active, data, dlen);
+    } else {
+        assert(ctx->xc_flags & VMBUS_XACT_CTXF_DESTROY); //no active xact pending
+        vmbus_xact_debug("vmbus: drop xact response\n");
```
Comment (francescolavra) [inline src/hyperv/vmbus/vmbus_xact.c:370]: '\n' is not needed

File: src/hyperv/storvsc/storvsc.c [hunk 2]
```diff
@@ -0,0 +1,1264 @@
...
+            &request->prp_list.gpa_range, request->prp_cnt,
+            vstor_packet, VSTOR_PKT_SIZE, (uint64_t)request);
+    } else {
+        ret = vmbus_chan_send(sc->hs_chan,
+            VMBUS_CHANPKT_TYPE_INBAND, VMBUS_CHANPKT_FLAG_RC,
+            vstor_packet, VSTOR_PKT_SIZE, (uint64_t)request);
+    }
+
+    if (ret != 0) {
+        rprintf("Unable to send packet %p ret %d", vstor_packet, ret);
```
Comment (francescolavra) [inline src/hyperv/storvsc/storvsc.c:523]: Suggest adding trailing '\n'.

File: src/hyperv/storvsc/storvsc.c [hunk 3]
```diff
@@ -0,0 +1,1264 @@
...
+ * Host control blocks (HCBs) are converted into VSCSI protocol messages which
+ * are delivered to the parent partition StorVSP driver over the Hyper-V VMBUS.
+ */
+
+#include <kernel.h>
+#include <page.h>
+#include <hyperv_internal.h>
+#include <hyperv.h>
+#include <vmbus.h>
+#include <net_system_structs.h>
```
Comment (francescolavra) [inline src/hyperv/storvsc/storvsc.c:42]: I guess this header inclusion is not needed.

File: src/hyperv/netvsc/hv_net_vsc.c [hunk 4]
```diff
@@ -0,0 +1,1021 @@
...
+    vmbus_chan_gpadl_connect(
+        device->channel, rx_buf_paddr,
+        net_dev->rx_buf_size, &net_dev->rx_buf_gpadl_handle);
+
+    /* sema_wait(&ext->channel_init_sema); KYS CHECK */
+
+    /* Notify the NetVsp of the gpadl handle */
+    init_pkt = &net_dev->channel_init_packet;
+
+    memset(init_pkt, 0, sizeof(nvsp_msg));
```
[the remark this answers is not in the document]
Comment (mkhon) [reply src/hyperv/netvsc/hv_net_vsc.c]: Fixed

File: src/hyperv/netvsc/hv_net_vsc.c [hunk 5]
```diff
@@ -0,0 +1,1021 @@
...
+     */
+    if (nvsp_vers >= NVSP_PROTOCOL_VERSION_2)
+        ret = hv_nv_send_ndis_config(device, ifp->mtu);
+
+    /*
+     * Send the NDIS version
+     */
+    init_pkt = &net_dev->channel_init_packet;
+
+    memset(init_pkt, 0, sizeof(nvsp_msg));
```
[the remark this answers is not in the document]
Comment (mkhon) [reply src/hyperv/netvsc/hv_net_vsc.c]: Fixed

File: src/hyperv/netvsc/hv_net_vsc.c [hunk 6]
```diff
@@ -0,0 +1,1021 @@
...
+    do {
+        struct vmbus_chanpkt_hdr *desc = (struct vmbus_chanpkt_hdr *)buffer;
+        int bytes_rxed = bufferlen;
+        ret = vmbus_chan_recv_pkt(device->channel,
+            desc, &bytes_rxed);
+        if (ret == ENOBUFS) {
+            /* Handle large packet */
+            deallocate(device->device->general, buffer, bufferlen);
+            buffer = allocate(device->device->general, bytes_rxed);
+            if (buffer == NULL) {
```
[the remark this answers is not in the document]
Comment (mkhon) [reply src/hyperv/netvsc/hv_net_vsc.c]: Fixed

File: src/hyperv/netvsc/hv_net_vsc.c [hunk 7]
```diff
@@ -0,0 +1,1021 @@
...
+        nvsp_vers = NVSP_PROTOCOL_VERSION_1;
+        ret = hv_nv_negotiate_nvsp_protocol(device, net_dev, nvsp_vers);
+        if (ret != 0) {
+            /* NVSP v1 failed, return bad status */
+            return (ret);
+        }
+    }
+    net_dev->nvsp_version = nvsp_vers;
+
+    netvsc_debug("NVSP protocol: %d\n", net_dev->nvsp_version);
```
[the remark this answers is not in the document]
Comment (mkhon) [reply src/hyperv/netvsc/hv_net_vsc.c]: Fixed

File: src/hyperv/vmbus/vmbus_chan.c [hunk 8]
```diff
@@ -0,0 +1,974 @@
...
+#include "hyperv_var.h"
+#include "vmbus_reg.h"
+#include "vmbus_var.h"
+#include "vmbus_brvar.h"
+#include "vmbus_chanvar.h"
+
+#define NULL 0
+
+#ifdef VMBUS_CHAN_DEBUG
+#define vmbus_chan_debug(x, ...) do {rprintf(" VMBUS CHAN: " x "\n", ##__VA_ARGS__);} while(0)
```
[the remark this answers is not in the document]
Comment (mkhon) [reply src/hyperv/vmbus/vmbus_chan.c:44]: Fixed

File: src/hyperv/storvsc/storvsc.c (hunk 2 above)
Comment (mkhon) [reply src/hyperv/storvsc/storvsc.c:523]: Fixed

File: src/hyperv/vmbus/vmbus_xact.c (hunk 1 above)
Comment (mkhon) [reply src/hyperv/vmbus/vmbus_xact.c]: Fixed

File: src/hyperv/vmbus/vmbus.c [hunk 9]
```diff
@@ -0,0 +1,658 @@
...
+     * checking messages.
+     */
+    sc->vmbus_event_proc(sc, cpu);
+
+    /*
+     * Check messages.  Mainly management stuffs; ultra low rate.
+     */
+    msg = msg_base + VMBUS_SINT_MESSAGE;
+    if (msg->msg_type != HYPERV_MSGTYPE_NONE) {
+        vmbus_debug("SINT Message!");
```
[the remark this answers is not in the document]
Comment (mkhon) [reply src/hyperv/vmbus/vmbus.c:98]: Fixed

File: src/hyperv/netvsc/hv_net_vsc.c [hunk 10]
```diff
@@ -0,0 +1,1021 @@
...
+    /*
+     * If we got a section count, it means we received a
+     * send_rx_buf_complete msg 
+     * (ie sent nvsp_msg_1_type_send_rx_buf msg) therefore,
+     * we need to send a revoke msg here
+     */
+    if (net_dev->send_section_size) {
+        /* Send the revoke send buffer */
+        revoke_pkt = &net_dev->revoke_packet;
+        memset(revoke_pkt, 0, sizeof(nvsp_msg));
```
[the remark this answers is not in the document]
Comment (mkhon) [reply src/hyperv/netvsc/hv_net_vsc.c]: Fixed

File: src/hyperv/netvsc/hv_rndis_filter.c [hunk 11]
```diff
@@ -0,0 +1,909 @@
...
+    case REMOTE_NDIS_RESET_CMPLT:
+    case REMOTE_NDIS_KEEPALIVE_CMPLT:
+        hv_rf_receive_response(rndis_dev, &rndis_mesg);
+        break;
+    /* notification message */
+    case REMOTE_NDIS_INDICATE_STATUS_MSG:
+        hv_rf_receive_indicate_status(rndis_dev, &rndis_mesg);
+        break;
+    default:
+        hyperv_rndis_debug("hv_rf_on_receive():  Unknown msg_type 0x%x\n",
```
[the remark this answers is not in the document]
Comment (mkhon) [reply src/hyperv/netvsc/hv_rndis_filter.c]: Fixed

File: src/hyperv/netvsc/hv_rndis_filter.c [hunk 12]
```diff
@@ -0,0 +1,909 @@
...
+        if (!request->wait_flag)
+            break;
+        kernel_delay(100);
+    }
+}
+
+static inline u64
+hv_get_phys_addr(void *virt)
+{
+    u64 ret = (physical_from_virtual(virt) | ((vm_offset_t) virt & PAGEMASK));
```
[the remark this answers is not in the document]
Comment (mkhon) [reply src/hyperv/netvsc/hv_rndis_filter.c]: Fixed

File: src/hyperv/netvsc/netvsc.c [hunk 13]
```diff
@@ -0,0 +1,425 @@
...
+    if (ret != 0)
+        return timm("err", "err");
+
+    netif_add(hn->netif,
+              0, 0, 0,
+              hn,
+              vmxif_init,
+              ethernet_input);
+
+    netvsc_debug("%s: hwaddr %02x:%02x:%02x:%02x:%02x:%02x\n", __func__,
```
[the remark this answers is not in the document]
Comment (mkhon) [reply src/hyperv/netvsc/netvsc.c]: Fixed

File: src/hyperv/vmbus/hyperv.c [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+        MSR_HV_HYPERCALL_ENABLE;
+    wrmsr(MSR_HV_HYPERCALL, hc);
+
+    /*
+     * Confirm that Hypercall page did get setup.
+     */
+    hc = rdmsr(MSR_HV_HYPERCALL);
+    if ((hc & MSR_HV_HYPERCALL_ENABLE) == 0) {
+        rprintf("hyperv: Hypercall setup failed\n");
+        halt(false);
```
[the remark this answers is not in the document]
Comment (mkhon) [reply src/hyperv/vmbus/hyperv.c]: Fixed

File: src/hyperv/netvsc/netvsc.c [hunk 15]
```diff
@@ -0,0 +1,425 @@
...
+        packet->compl.send.send_completion_tid = 0;
+
+        /*
+         * Release the resources since we will not get any
+         * send completion
+         */
+        netvsc_xmit_completion(packet);
+
+        // TODO - error code?
+        netvsc_debug("%s: hv_rf_on_send() failed %d times, giving up\n", __func__, retries);
```
[the remark this answers is not in the document]
Comment (mkhon) [reply src/hyperv/netvsc/netvsc.c]: Fixed

File: src/hyperv/netvsc/hv_rndis_filter.c [hunk 16]
```diff
@@ -0,0 +1,909 @@
...
+    query->oid = oid;
+    query->info_buffer_offset = sizeof(rndis_query_request); 
+    query->info_buffer_length = 0;
+    query->device_vc_handle = 0;
+
+    hv_request_prepare_wait(request);
+    ret = hv_rf_send_request(device, request, REMOTE_NDIS_QUERY_MSG);
+    if (ret != 0) {
+        /* Fixme:  printf added */
+        hyperv_rndis_debug("RNDISFILTER request failed to Send!\n");
```
[the remark this answers is not in the document]
Comment (mkhon) [reply src/hyperv/netvsc/hv_rndis_filter.c]: Fixed

File: src/hyperv/storvsc/storvsc.c (hunk 3 above)
Comment (mkhon) [reply src/hyperv/storvsc/storvsc.c:42]: It is needed because of ENOBUFS definition

**Comment by mkhon** [comment]
@eyberg assert when built without any DEBUG is now fixed

Network driver stalls is still an issue - I would suggest to fix them in a separate PR as this one is too big and waits for an approval for too long.

File: src/x86_64/service.c [hunk 17]
```diff
@@ -398,9 +399,13 @@ static void __attribute__((noinline)) init_service_new_stack()
     if (!kvm_detect(kh)) {
         init_debug("probing for Xen hypervisor");
         if (!xen_detect(kh)) {
-            init_debug("neither KVM nor Xen detected; assuming qemu full emulation");
-            if (!init_hpet(kh)) {
-                halt("HPET initialization failed; no timer source\n");
+            if (!hyperv_detect(kh)) {
+                init_debug("neither KVM nor Xen detected; assuming qemu full emulation");
```
Comment (wjhun) [inline src/x86_64/service.c:403]: Update text here ... e.g. "no hypervisor detected"

File: src/hyperv/vmbus/vmbus.c [hunk 18]
```diff
@@ -0,0 +1,662 @@
...
+
+void
+vmbus_handle_intr(vmbus_dev dev)
+{
+    int cpu = 0;
+
+    /*
+     * Disable preemption.
+     */
+    u64 flags = read_flags();
```
Comment (wjhun) [inline src/hyperv/vmbus/vmbus.c:111]: nit: can use irq_disable_save here

File: src/hyperv/vmbus/hyperv_var.h [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * $FreeBSD$
+ */
+
+#ifndef _HYPERV_VAR_H_
+#define _HYPERV_VAR_H_
+
+static inline u64 rdmsr(u32 idx)
```
Comment (wjhun) [inline src/hyperv/vmbus/hyperv_var.h:32]: we have write_msr and read_msr

**Comment by eyberg** [comment]
yeh - i'd just fix these comments, update to master and then merge - can do the network issue in a sep. pr

File: src/x86_64/service.c (hunk 17 above)
Comment (mkhon) [reply src/x86_64/service.c]: Fixed

File: src/hyperv/vmbus/hyperv_var.h (hunk 19 above)
Comment (mkhon) [reply src/hyperv/vmbus/hyperv_var.h]: Fixed

**Pull request merged**